libraries/app.rb

```diff
@@ -0,0 +1,149 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Cookbook Name:: xcode
+# Library:: app
+#
+require 'chef/resource'
+require 'chef/provider'
+require 'chef/provider/ruby_block'
+
+module Xcode
+  module Resources
+    module XcodeApp
+      # The 'xcode_app' custom resource to install and uninstall Xcode components.
+      class Resource < Chef::Resource::LWRPBase
+        self.resource_name = :xcode_app
+        provides(:xcode_app)
+        actions(:install, :uninstall)
+        default_action(:install)
+
+        attribute(:version, kind_of: String, name_attribute: true)
+        attribute(:url, kind_of: String, required: true)
+        attribute(:checksum, kind_of: String, required: true)
+        attribute(:app, kind_of: String, required: true)
+        attribute(:multi_install, kind_of: [TrueClass, FalseClass], default: true)
+        attribute(:install_root, kind_of: String, default: '/Applications')
+        attribute(:force, kind_of: [TrueClass, FalseClass], default: false)
+      end
+
+      # The 'xcode_app' custom resource provider to install and uninstall Xcode components.
+      class Provider < Chef::Provider::LWRPBase
+        provides(:xcode_app)
+        use_inline_resources
+
+        action :install do
+          return if exist?
+          raise unless Chef.node.platform_family.eql?('mac_os_x')
+          install_app
+        end
+
+        action :uninstall do
+          directory install_dir do
+            recursive true
+            action :delete
+          end if exist?
+        end
+
+        def install_app
+          # If we force the install remove target folder if it exists.
+          directory install_dir do
+            recursive true
+            action :delete
+          end if new_resource.force && exist?
+
+          if new_resource.url.end_with?('dmg')
+            install_dmg
+          else
+            raise 'Unsupported package provided Xcode installer must be provided as a DMG'
+          end
+
+          post_install
+          accept_eula
+          delete_installer_file if exist?
+        end
+
+        def install_dmg
+          dmg_package new_resource.app do
+            source new_resource.url
+            checksum new_resource.checksum
+            dmg_name ::File.basename(new_resource.url, '.dmg')
+            owner 'root'
+            type 'app'
+            action :install
+          end
+        end
+
+        # The `post_install` function is only called if the function 'install_app' is called.
+        def post_install
+          ruby_block 'Xcode post install' do
+            block do
+              if new_resource.multi_install || !new_resource.install_root.eql?('/Applications')
+                ::File.rename('/Applications/Xcode.app', install_dir)
```

Comment:
Why move/rename as opposed to just setting `destination` attribute when calling `dmg_package`? Saves a step.